op_crates/http/lib.rs

```diff
@@ -0,0 +1,457 @@
+use deno_core::serde_json::Value;
+use deno_core::AsyncRefCell;
+use deno_core::BufVec;
+use deno_core::CancelFuture;
+use deno_core::CancelHandle;
+use deno_core::CancelTryFuture;
+use deno_core::JsRuntime;
+use deno_core::OpState;
+use deno_core::RcRef;
+use deno_core::Resource;
+use deno_core::ZeroCopyBuf;
+
+use hyper::body::HttpBody;
+use hyper::http;
+use hyper::server::conn::Connection;
+use hyper::server::conn::Http;
+use hyper::service::Service;
+use hyper::Body;
+use hyper::Request;
+use hyper::Response;
+use serde::Deserialize;
+use tokio::io::AsyncReadExt;
+use tokio::net::TcpListener;
+use tokio::sync::oneshot;
+use tokio_util::io::StreamReader;
+
+/// Load and execute the javascript code.
+pub fn init(isolate: &mut JsRuntime) {
+  let files =
+    vec![("deno:op_crates/http/00_http.js", include_str!("00_http.js"))];
+  for (url, source_code) in files {
+    isolate.execute(url, source_code).unwrap();
+  }
+}
+
+pub fn get_declaration() -> PathBuf {
+  PathBuf::from(env!("CARGO_MANIFEST_DIR")).join("lib.deno_http.d.ts")
+}
+
+// Needed so hyper can use non Send futures
+#[derive(Clone)]
+struct LocalExecutor;
+
+impl<Fut> hyper::rt::Executor<Fut> for LocalExecutor
+where
+  Fut: Future + 'static,
+  Fut::Output: 'static,
+{
+  fn execute(&self, fut: Fut) {
+    tokio::task::spawn_local(fut);
+  }
+}
+
+struct DenoServiceInner {
+  pub request: Request<Body>,
+  pub response_tx: oneshot::Sender<Response<Body>>,
+}
+
+#[derive(Clone)]
+struct DenoService {
+  inner: Rc<RefCell<Option<DenoServiceInner>>>,
+}
+
+impl Default for DenoService {
+  fn default() -> Self {
+    Self {
+      inner: Rc::new(RefCell::new(None)),
+    }
+  }
+}
+
+impl Service<Request<Body>> for DenoService {
+  type Response = Response<Body>;
+  type Error = http::Error;
+  #[allow(clippy::type_complexity)]
+  type Future =
+    Pin<Box<dyn Future<Output = Result<Self::Response, Self::Error>>>>;
+
+  fn poll_ready(
+    &mut self,
+    _cx: &mut Context<'_>,
+  ) -> Poll<Result<(), Self::Error>> {
+    Poll::Ready(Ok(()))
+  }
+
+  fn call(&mut self, req: Request<Body>) -> Self::Future {
+    let (resp_tx, resp_rx) = oneshot::channel();
+    self.inner.borrow_mut().replace(DenoServiceInner {
+      request: req,
+      response_tx: resp_tx,
+    });
+
+    async move { Ok(resp_rx.await.unwrap()) }.boxed_local()
+  }
+}
+
+struct HttpServer {
+  pub listener: TcpListener,
+}
+
+impl Resource for HttpServer {
+  fn name(&self) -> Cow<str> {
+    "httpServer".into()
+  }
+}
+
+struct ConnResource {
+  pub hyper_connection:
+    Rc<RefCell<Connection<tokio::net::TcpStream, DenoService, LocalExecutor>>>,
+  pub deno_service: DenoService,
+}
+
+impl Resource for ConnResource {
+  fn name(&self) -> Cow<str> {
+    "httpConnection".into()
+  }
+}
+
+#[derive(Deserialize)]
+pub struct CreateServerArgs {
+  address: String,
+}
+
+pub async fn op_create_server(
+  state: Rc<RefCell<OpState>>,
+  args: CreateServerArgs,
+  _data: BufVec,
+) -> Result<Value, AnyError> {
```

Comment:
This op should check net permission

Reply:
Sure, sounds good

Reply:
Actually not, it's a separate crate and doesn't have access to `TcpListenerResource`

Reply:
Mh - that is very unfortunate...

Reply:
Wasn't planning on doing that in the first pass

Reply:
Ok - sounds like we will need to move our net ops into a seperate crate then soon (or move this into `runtime/ops/`?). I think the http crate should _upgrade_ a connection and not duplicate starting of listeners. Anyway not so important for first pass.